### PR TITLE
Fix to #6576 - Query: issue warning for queries with paging operations (First, Take, Skip) without OrderBy

### PR DESCRIPTION
**Pull request by maumar**
Performing paging operations without OrderBy may lead to unpredictable results and could be hard to track down. Fix is to issue a warning in the logger if such situation happens.

### CODE REVIEW

File: src/Microsoft.EntityFrameworkCore/Query/EntityQueryModelVisitor.cs [hunk 1]
```diff
@@ -659,6 +659,14 @@ public override void VisitQueryModel([NotNull] QueryModel queryModel)
         {
             Check.NotNull(queryModel, nameof(queryModel));
 
+            if (queryModel.ResultOperators.Any(o => o is FirstResultOperator || o is SkipResultOperator || o is TakeResultOperator)
+                && !queryModel.BodyClauses.OfType<OrderByClause>().Any())
+            {
+                QueryCompilationContext.Logger.LogWarning(
+                    CoreEventId.CompilingQueryModel,
```
Comment (maumar) [inline src/Microsoft.EntityFrameworkCore/Query/EntityQueryModelVisitor.cs]: should events be more granular?

File: src/Microsoft.EntityFrameworkCore/Query/EntityQueryModelVisitor.cs [hunk 2]
```diff
@@ -659,6 +659,14 @@ public override void VisitQueryModel([NotNull] QueryModel queryModel)
...
+                    () => CoreStrings.PagingOperationWithoutOrderBy(queryModel.MainFromClause.ItemType.Name));
```
Comment (maumar) [inline src/Microsoft.EntityFrameworkCore/Query/EntityQueryModelVisitor.cs]: using MainFromClause is not always perfect - in some cases the type can be compiler generated which doesn't tell much to the user. Are we ok with that?

**Comment by maumar** [comment]
@anpete @divega @rowanmiller

File: src/Microsoft.EntityFrameworkCore/Query/EntityQueryModelVisitor.cs (hunk 1 above)
Comment (anpete) [inline src/Microsoft.EntityFrameworkCore/Query/EntityQueryModelVisitor.cs]: I think this is fine.

File: src/Microsoft.EntityFrameworkCore.Specification.Tests/QueryTestBase.cs [hunk 3]
```diff
@@ -6371,6 +6371,34 @@ public virtual void DefaultIfEmpty_in_subquery_nested()
...
+        public virtual void Paging_operation_without_orderby_issues_warning_subquery()
+        {
+            using (var context = CreateContext())
+            {
+                context.Customers.Where(c => c.CustomerID == "ALFKI" && c.Orders.FirstOrDefault().OrderID > 1000).ToList();
+            }
+        }
+
+        [ConditionalFact]
+        public virtual void Single_SingleOrDefault_without_orderby_doesnt_issue_warning()
```
Comment (maumar) [inline src/Microsoft.EntityFrameworkCore.Specification.Tests/QueryTestBase.cs]: single will throw if there is more than 1 result and if its one result then its deterministic already

File: src/Microsoft.EntityFrameworkCore/Query/EntityQueryModelVisitor.cs (hunk 2 above)
Comment (maumar) [inline src/Microsoft.EntityFrameworkCore/Query/EntityQueryModelVisitor.cs]: I tried that initially but it looked really bloated for a bit more complicated query. @divega was proposing that we could assign an alias to each query model and refer to them via alias in all the log messages. If we opted for it, it would be in different pull request.

File: src/Microsoft.EntityFrameworkCore/Properties/CoreStrings.resx [hunk 4]
```diff
@@ -657,4 +657,7 @@
   <data name="PropertyCalledOnNavigation" xml:space="preserve">
     <value>Cannot call Property for the property '{property}' on entity type '{entityType}' because it is configured as a navigation property. Property can only be used to configure scalar properties.</value>
   </data>
+  <data name="PagingOperationWithoutOrderBy" xml:space="preserve">
+    <value>Query on '{querySourceType}' uses paging operation (First/FirstOrDefault/Skip/Take) without OrderBy which may lead to unpredictable results.</value>
```
Comment (maumar) [inline src/Microsoft.EntityFrameworkCore/Properties/CoreStrings.resx]: good idea, will do

File: src/Microsoft.EntityFrameworkCore.Specification.Tests/QueryTestBase.cs (hunk 3 above)
Comment (anpete) [inline src/Microsoft.EntityFrameworkCore.Specification.Tests/QueryTestBase.cs]: Makes sense.

File: src/Microsoft.EntityFrameworkCore/Query/EntityQueryModelVisitor.cs (hunk 2 above)
Comment (maumar) [inline src/Microsoft.EntityFrameworkCore/Query/EntityQueryModelVisitor.cs]: sounds like a reasonable compromise

File: src/Microsoft.EntityFrameworkCore/Properties/CoreStrings.resx (hunk 4 above)
Comment (maumar) [inline src/Microsoft.EntityFrameworkCore/Properties/CoreStrings.resx]: one problem is for cases like:
```
customers.Where(c => c.Orders.First().Id == 1)
```

This doesn't have a filter, but we add one during nav rewrite, so we won't issue warning for this case, even though we should

File: src/Microsoft.EntityFrameworkCore/Query/EntityQueryModelVisitor.cs (hunk 1 above)
Comment (anpete) [inline src/Microsoft.EntityFrameworkCore/Query/EntityQueryModelVisitor.cs]: Consider extracting this into a protected virtual method.

File: src/Microsoft.EntityFrameworkCore/Query/EntityQueryModelVisitor.cs (hunk 2 above)
Comment (anpete) [inline src/Microsoft.EntityFrameworkCore/Query/EntityQueryModelVisitor.cs]: Cool, we should make sure there are no newlines, too. We could just use QM.ToString(), which is single line formatted.

File: src/Microsoft.EntityFrameworkCore/Properties/CoreStrings.resx (hunk 4 above)
Comment (anpete) [inline src/Microsoft.EntityFrameworkCore/Properties/CoreStrings.resx]: Do it before nav. rewrite?

File: src/Microsoft.EntityFrameworkCore/Query/EntityQueryModelVisitor.cs (hunk 2 above)
Comment (maumar) [inline src/Microsoft.EntityFrameworkCore/Query/EntityQueryModelVisitor.cs]: im adding no-formatting option to expression/qm printers. default tostring leaves full type names which "steal" way to much space sometimes

**Comment by maumar** [comment]
@anpete @divega  new version up

File: src/Microsoft.EntityFrameworkCore/Query/EntityQueryModelVisitor.cs [hunk 5]
```diff
@@ -360,6 +364,39 @@ var entityEqualityRewritingExpressionVisitor
...
+            }
+
+            public override void VisitQueryModel(QueryModel queryModel)
+            {
+                if (queryModel.ResultOperators.Any(o => o is SkipResultOperator || o is TakeResultOperator)
+                    && !queryModel.BodyClauses.OfType<OrderByClause>().Any())
+                {
+                    _logger.LogWarning(
+                        CoreEventId.CompilingQueryModel,
+                        () => CoreStrings.RowLimitingOperationWithoutOrderBy(queryModel.Print(true, QueryModelStringLengthLimit)));
```
Comment (anpete) [inline src/Microsoft.EntityFrameworkCore/Query/EntityQueryModelVisitor.cs]: Add Parameter name in Print

File: test/Microsoft.EntityFrameworkCore.SqlServer.FunctionalTests/QueryLoggingSqlServerTest.cs [hunk 6]
```diff
@@ -30,10 +30,10 @@ var customers
                 Assert.NotNull(customers);
                 Assert.StartsWith(
                     @"    Compiling query model: 
-'from Customer <generated>_0 in DbSet<Customer>
+'from Customer <generated>_0 in DbSet<Customer>
```
Comment (anpete) [inline test/Microsoft.EntityFrameworkCore.SqlServer.FunctionalTests/QueryLoggingSqlServerTest.cs]: This extra space expected?

File: src/Microsoft.EntityFrameworkCore/Query/ExpressionVisitors/Internal/RecursiveQueryModelVisitorBase.cs [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+using System.Linq.Expressions;
+using Remotion.Linq.Clauses.Expressions;
+
+namespace Microsoft.EntityFrameworkCore.Query.ExpressionVisitors.Internal
+{
+    /// <summary>
+    ///     This API supports the Entity Framework Core infrastructure and is not intended to be used
+    ///     directly from your code. This API may change or be removed in future releases.
+    /// </summary>
+    public abstract class RecursiveQueryModelVisitorBase : ExpressionTransformingQueryModelVisitor<ExpressionVisitorBase>
```
Comment (anpete) [inline src/Microsoft.EntityFrameworkCore/Query/ExpressionVisitors/Internal/RecursiveQueryModelVisitorBase.cs]: Going to try and unify to TransformExpressions here

File: test/Microsoft.EntityFrameworkCore.SqlServer.FunctionalTests/QueryLoggingSqlServerTest.cs (hunk 6 above)
Comment (anpete) [inline test/Microsoft.EntityFrameworkCore.SqlServer.FunctionalTests/QueryLoggingSqlServerTest.cs]: Is it hard to fix this extra space?

**Comment by anpete** [comment]
:shipit:
